Refactored and added new commands

whiterock opened the pull request:
Refactored the keypress detection such that we can easily detect stuff like `gg` and `yy`. Its prettier and more flexible this way. My testing showed that everything still works as intended as before, verify for yourself :) Note: I am not a js native!

whiterock committed:
Refactored and added new commands

Refactored the keypress detection such that we can easily detect stuff like `gg` and `yy`. Its prettier and more flexible this way. My testing showed that everything still works as intended as before, verify for yourself :) Note: I am not a js native!

whiterock committed:
Updated README.md to reflect additions

PalmerAL commented:
This is pretty nice!

The only thing I have a question about is the close tab shortcut. There's already a customizable keybinding in the Min preferences to close a tab. It doesn't support multiple letters; however, looking at the keybindings for [Vimium](https://vimium.github.io/), that extension is using `x` rather than `dd`. Unless there's a reason to use `dd`, I'm wondering if it would make more sense to use `x`, and if a single-letter shortcut is fine, I'm wondering if it would be better to just let people customize this from the settings page rather than adding an additional shortcut. Thoughts?

whiterock committed:
removed `dd` as suggested

whiterock commented:
Yes, you're right, I removed it!

whiterock committed:
improved follow link logic

it now uses a flexible alphabet, i.e. we can exclude keys that collide with min keybindings set and the lettering algo is a bit different in that it starts with the first letter in the alphabet for the second letter for double letter combinations now.

whiterock commented:
I also changed the way the follow link letters are generated. It can now use an arbitrary alphabet, which has the advantage that we can exclude letters set in the min preferences. For example, I have `h` and `l` set for tab switching and `x` set for tab closing as we discussed in the previous post, thus I can exclude these letters such that I don't switch tabs when following the link `nh` e.g.

PalmerAL commented on README.md:
```diff
@@ -5,6 +5,10 @@ Adds Vim-style shortcuts to [Min](https://github.com/minbrowser/min):
+* Press gg to scroll to the top
+* Press G to scroll to the bottom
+* Press yy to copy the current url
+* Press dd to close the tab
```
can you remove this?

whiterock replied:
removed!

PalmerAL replied:
Actually, I didn't mean to remove all of them, just the "dd" shortcut since that isn't included anymore.

whiterock replied:
ah my bad, fixed.

PalmerAL commented on global.js:
```diff
@@ -1,4 +1,12 @@
+let problematic_letters = 'hlx'.split('')
+// This a set-theoretic difference, removing the unwanted letters
+var alphabet = 'abcdefghijklmnopqrstuvwxyz'.split('').filter(x => !problematic_letters.includes(x));
+console.log(alphabet)
```
Remove this as well.

PalmerAL commented on global.js:
```diff
@@ -23,11 +31,15 @@ function isVisible (rect) {
 
 function getNextKeyCombination (index) {
   // use 1st half of alphabet for single letter, and 2nd half for double letter, so that no collisions can occur
-  if (index < 13) {
+  let halfIndex = Math.floor(alphabet.length / 2)
+  if (index < halfIndex) {
     return alphabet[index]
   } else {
+    // Subtract half-index from index such that lettering starts at first index for second letter, e.g.
+    // na nb nc nd ...
+    index -= halfIndex
     // two-letter combination
-    return alphabet[Math.floor(index / 13) + 12] + alphabet[index % 26]
+    return alphabet[Math.floor(index / alphabet.length) + halfIndex] + alphabet[index % alphabet.length]
```
This seems different than the original logic - the original code goes `m`, `nn`, `no`, whereas this does `m`, `na`, `nb`. I don't actually remember at the moment what the original purpose of splitting the alphabet was though, so if you think the new way is fine then I'm good with it.

whiterock replied:
I think the new way is fine indeed.

PalmerAL commented:
I'm fine with the approach of excluding letters for now, but it seems like a better approach might be to prevent the single-character shortcuts from running when in link following mode. Single-character shortcuts will already be ignored whenever you're typing in an input, so I think it would probably work to add an invisible input element when entering link-following mode, and have the letters be typed into that. That might fix the GitHub issue as well.

It's fine to leave it the way it is though if you want.

whiterock committed:
removed lines from readme as suggested

whiterock commented:
> I'm fine with the approach of excluding letters for now, but it seems like a better approach might be to prevent the single-character shortcuts from running when in link following mode. Single-character shortcuts will already be ignored whenever you're typing in an input, so I think it would probably work to add an invisible input element when entering link-following mode, and have the letters be typed into that. That might fix the GitHub issue as well.
> 
> It's fine to leave it the way it is though if you want.

I tried to implement this for the last 2 hours now, it kinda works but for some reason it's really buggy and just randomly works or doesn't. For example if I follow the link created called `j` there is about a 50/50 chance it will change tabs one to the left after opening the link in the same tab. I have currently no clue as to why this is happening and my debugging hasn't gone anywhere either. I am particulary confused as this is a phenomenon I have never had to deal with in any programming language but I will be sure to look into it as soon as I got the nerves and time on hand. You can check out what I did if you'd like, I put it in a separate branch called experimental: https://github.com/whiterock/min-vim-mode/blob/experimental/global.js

PalmerAL commented:
What keyboard shortcuts do you have configured in the preferences, and what keyboard layout do you have?

whiterock committed:
fix readme

whiterock commented:
> What keyboard shortcuts do you have configured in the preferences, and what keyboard layout do you have?

<img width="657" alt="Screenshot 2020-05-08 at 23 34 45" src="https://user-images.githubusercontent.com/6281495/81451335-903f4e80-9184-11ea-908c-d5954384ca12.png">
The keyboard shortcuts not shown here are the default ones.

I use a german-style QWERTZ keyboard.

whiterock commented:
I have found out something just now: Pressing `f` and `h` doesn't result in "change tab one left" if I do it in fast enough succession, i.e. it properly follows the link with the h rectangle and doesn't change tabs afterwards. Now this a good observation but it still leaves me clueless as to why that would affect my code since it doesn't rely on any timers/delays and things of the sort.

PalmerAL commented:
I can reproduce this if I hold down the `h` key for a second longer than I normally would; that might be what you're seeing, particularly if you've adjusted the "delay until repeat" OS setting.

The reason for that is that the script is using `keypress`, which will fire while the key is still being held down. When you press the key and the event fires, the script will navigate to the new page, but the input will be gone as soon as the navigation starts. If you keep holding the key down, the browser UI will receive another keypress event, and since there isn't any input focused any more, it will run the default shortcut handler, which is to switch tabs.

Using `keyup` in the script instead should fix this.

whiterock committed:
Fixed the bug where keypresses would propagate to the browser

aided by the suggestion of using keyup

whiterock commented:
> I can reproduce this if I hold down the `h` key for a second longer than I normally would; that might be what you're seeing, particularly if you've adjusted the "delay until repeat" OS setting.
> 
> The reason for that is that the script is using `keypress`, which will fire while the key is still being held down. When you press the key and the event fires, the script will navigate to the new page, but the input will be gone as soon as the navigation starts. If you keep holding the key down, the browser UI will receive another keypress event, and since there isn't any input focused any more, it will run the default shortcut handler, which is to switch tabs.
> 
> Using `keyup` in the script instead should fix this.

Thank you for this insight. I was able to fix it now and in all of my testing it worked flawlessly and better yet, it's a pleasure to use. I'd love it if you could merge this now :)

PalmerAL committed:
A few additional fixes

* Reformat with standard-format
* Set the vertical position of the input so that typing into it doesn't cause the page to scroll
* use addEventListener to avoid overwriting an existing listener created by the page

PalmerAL commented:
I fixed a couple minor things, but I think everything else looks good. It does work pretty well - thanks for working on this!

whiterock commented:
Thanks, just learned some new js stuff by looking a the fixes you did!

No problem, I will be sure to add some more features when I find they‘re missing and I have some time :)
